### PR TITLE
Use conda to build python packages during GPU tests

### PR DESCRIPTION
**Pull request by Ethyling**
This PR convert the `from sources` build we are doing in GPU test job to a `conda build`. This is done for the following reasons:
- This is required step to improve the Ops CI/CD setup to a more convenient pipeline
- This is required to start using `conda` compilers and `mamba` to build RAPIDS packages
- This prevent us from manually managing and installing the dependencies in GPU job
- This ensure the packages can be installed
- This ensure the tests are running and working against the package content and not the build results. Currently the Python packages are not tested.

This may increase the global pipeline time, but the usage of `mamba` should resolve this as `mamba` is faster than `conda` to build packages

### CODE REVIEW

**Comment by pentschev** [comment]
Is there some advantage to doing this, or is it just to test conda packaging as well?

**Comment by Ethyling** [comment]
> Is there some advantage to doing this, or is it just to test conda packaging as well?
That's one of the advantages. We are also doing this to remove the usage of the `rapids-build-env` package and to start using conda compilers with mamba to build packages. I will update the PR body with all of this before merging it

**Comment by Ethyling** [comment]
rerun tests

**Comment by Ethyling** [comment]
rerun tests

**Comment by pentschev** [comment]
rerun tests

**Comment by pentschev** [comment]
The timeout from `dask_cuda.tests.test_spill.test_cudf_cluster_device_spill[params0]` is being dealt with in https://github.com/rapidsai/dask-cuda/pull/901 .

**Comment by pentschev** [comment]
rerun tests

**Comment by pentschev** [comment]
rerun tests

**Comment by pentschev** [comment]
This many timeouts doesn't look normal, even the timeout that was increased in https://github.com/rapidsai/dask-cuda/pull/901 was hit here. @Ethyling could you add `conda list --show-channels` after `gpuci_mamba_retry install dask-cuda` so I can compare the packages being picked here vs what was being picked before?

**Comment by Ethyling** [comment]
> This many timeouts doesn't look normal, even the timeout that was increased in #901 was hit here. @Ethyling could you add `conda list --show-channels` after `gpuci_mamba_retry install dask-cuda` so I can compare the packages being picked here vs what was being picked before?

@pentschev PR updated with this

**Comment by pentschev** [comment]
rerun tests

**Comment by pentschev** [comment]
rerun tests

**Comment by pentschev** [comment]
rerun tests

**Comment by pentschev** [comment]
rerun tests

**Comment by pentschev** [comment]
Three successful runs in a row. At this point I would assume we had a hiccup in CI when this was running before and would suggest we move forward as is.

@Ethyling is there anything more that needs to be done here or should we consider reviewing/merging this PR?

**Comment by Ethyling** [comment]
> Three successful runs in a row. At this point I would assume we had a hiccup in CI when this was running before and would suggest we move forward as is.

Nice, thank you @pentschev!

> @Ethyling is there anything more that needs to be done here or should we consider reviewing/merging this PR?

I prefer to wait that all the PRs pass for all the RAPIDS projects before merging it

File: ci/gpu/build.sh
Comment (ajschmidt8) [inline ci/gpu/build.sh:83]: quick question. was this block removed since `dask` is installed via `dask-cudf`?

**Review by ajschmidt8 — APPROVED** [review]
Approved with a question.

File: ci/gpu/build.sh
Comment (ajschmidt8) [inline ci/gpu/build.sh:80]: this needs a `-c` flag to ensure that `dask-cuda` is installed from the local filesystem

**Review by ajschmidt8 — CHANGES_REQUESTED** [review]

File: ci/gpu/build.sh
Comment (Ethyling) [reply ci/gpu/build.sh]: PR updated

File: conda/recipes/dask-cuda/meta.yaml
Comment (Ethyling) [inline conda/recipes/dask-cuda/meta.yaml:13]: This change is required to prevent this issue:
```
AssertionError: Can't merge/copy source into subdirectory of itself.  Please create separate spaces for these things.
  src: /workspace
  dst: /workspace/.conda-bld/work
```

File: ci/gpu/build.sh
Comment (pentschev) [reply ci/gpu/build.sh:83]: But we always want to install Dask from the main branch in CI, are we doing that via Dask conda nightly channel?

Comment (charlesbluca) [reply ci/gpu/build.sh:83]: Doesn't seem like it from the gpuCI runs - should we consider taking an approach similar to [cuDF's](https://github.com/rapidsai/cudf/blob/1519108adcdb520ff4b41849d7dc6cf61330c2fc/ci/gpu/build.sh#L92-L99) here?

Comment (Ethyling) [reply ci/gpu/build.sh:83]: This commit should solve this: https://github.com/rapidsai/dask-cuda/compare/b0024dbbbeaf659aac43118443a2aac4033e9615..601f9a79be30df766587886c748858413e26861e

Comment (pentschev) [reply ci/gpu/build.sh:83]: We would prefer to keep the `INSTALL_DASK_MAIN` environment variable as it was before, and as is in the cuDF CI script that Charles mentioned above. Every release during code freeze we just flip that variable to use the release version of Dask and then flip it over again on the new branch.

**Comment by pentschev** [comment]
Failing test is due to https://github.com/dask/distributed/issues/6320 , which was caused by https://github.com/dask/distributed/pull/5910 merge.

**Comment by Ethyling** [comment]
> Failing test is due to [dask/distributed#6320](https://github.com/dask/distributed/issues/6320) , which was caused by [dask/distributed#5910](https://github.com/dask/distributed/pull/5910) merge.

Do you know if this will be fixed soon?

**Comment by pentschev** [comment]
Sorry @Ethyling , I don't know if that is going to be fixed soon. I marked the test to xfail in https://github.com/rapidsai/dask-cuda/pull/908 to unblock CI for now.

**Comment by pentschev** [comment]
rerun tests

**Review by pentschev — APPROVED** [review]
LGTM, thanks for the work here @Ethyling .

**Comment by pentschev** [comment]
Feel free to merge if this is ready to go.

**Comment by Ethyling** [comment]
Thank you for your help with this PR @pentschev!

**Comment by Ethyling** [comment]
@gpucibot merge

**Pull request merged**
